lib/logstash/inputs/jms.rb

```diff
@@ -347,4 +418,101 @@ def get_root_cause(e)
     end
     {:exception => cause.class.name, :exception_message => cause.message }
   end
+
+  private
+
+  def normalize_field_ref(target)
+    # so we can later event.set("#{target}[#{name}]", ...)
+    target.match?(/\A[^\[\]]+\z/) ? "[#{target}]" : target
+  end
+
+  def event_setter_for(target)
+    if target.nil? || target.empty?
+      TOP_LEVEL_EVENT_SETTER
+    else
+      TargetEventSetter.new normalize_field_ref(target)
+    end
+  end
+
+  TOP_LEVEL_EVENT_SETTER = lambda { |event, data| data.each { |key, val| event.set(key, val) } }
+  private_constant :TOP_LEVEL_EVENT_SETTER
+
+  class TargetEventSetter
+
+    def initialize(target)
+      @target = target
+    end
+
+    def call(event, data)
+      data.each { |key, val| event.set("#{@target}[#{key}]", val) }
+    end
+
+  end
+  private_constant :TargetEventSetter
+
+  def map_headers(msg)
+    @headers_mapper.call(msg)
+  end
+
+  # Maps JMS headers names
+  # (similar to JMS gem's Message#attributes extension)
+  class HeadersMapper
+
+    # @param msg [javax.jms.Message]
+    # @return [Hash]
+    def call(msg)
+      map = {
+        'jms_message_id' => msg.getJMSMessageID, # String
+        'jms_timestamp' => msg.getJMSTimestamp, # long
+        'jms_expiration' => msg.getJMSExpiration, # long
+        'jms_priority' => msg.getJMSPriority, # int (0-9)
+        'jms_type' => msg.getJMSType, # String
+        'jms_redelivered' => msg.getJMSRedelivered, # boolean
+      }
+
+      correlation_id = msg.getJMSCorrelationID # String
+      map['jms_correlation_id'] = correlation_id unless correlation_id.nil?
+
+      delivery_mode = jms_delivery_mode(msg)
+      map['jms_delivery_mode'] = delivery_mode unless delivery_mode.nil?
+
+      destination = msg.getJMSDestination # javax.jms.Destination
+      map['jms_destination'] = destination.to_string unless destination.nil?
+
+      reply_to = msg.getJMSReplyTo # javax.jms.Destination
+      map['jms_reply_to'] = reply_to.to_string unless reply_to.nil?
+
+      map
+    end
+
+    private
+
+    def jms_delivery_mode(msg)
+      case msg.getJMSDeliveryMode
+      when javax.jms.DeliveryMode::PERSISTENT
+        'persistent'
+      when javax.jms.DeliveryMode::NON_PERSISTENT
+        'non_persistent'
+      else
+        nil
+      end
+    end
+
+    INSTANCE = self.new
+
+  end
+
+  # For plugin compatibility due the use of JMS gem's Message#attributes.
+  class LegacyHeadersMapper < HeadersMapper
+
+    def call(msg)
+      map = super(msg)
+      map['jms_delivery_mode_sym'] = jms_delivery_mode(msg)
```

Comment:
If we are looking to _rename_ to match the `Message#attributes` behaviour, then:

```suggestion
      delivery_mode_sym = map.delete('jms_delivery_mode')&.to_sym
      map['jms_delivery_mode_sym'] = delivery_mode_sym unless delivery_mode_sym.nil?
```

Reply:
as the headers weren't documented - was thinking the best path would be to also add `jms_delivery_mode` in legacy mode,
document the headers and keep `jms_delivery_mode_sym` due compatibility ... is this problematic that we provide an extra (duplicate) header in legacy mode?

Reply:
I don't think that the extra field is too problematic, and am okay shipping both.